src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs

```diff
@@ -256,13 +257,23 @@ private Document CreateLuceneDocument(DocumentIndex documentIndex)
                     case DocumentIndex.Types.Text:
                         if (entry.Value != null && !String.IsNullOrEmpty(Convert.ToString(entry.Value)))
                         {
+                            var stringValue = Convert.ToString(entry.Value);
+
                             if (entry.Options.HasFlag(DocumentIndexOptions.Analyze))
                             {
-                                doc.Add(new TextField(entry.Name, Convert.ToString(entry.Value), store));
+                                doc.Add(new TextField(entry.Name, stringValue, store));
                             }
                             else
                             {
-                                doc.Add(new StringField(entry.Name, Convert.ToString(entry.Value), store));
+                                doc.Add(new StringField(entry.Name, stringValue, store));
+                            }
+
+                            // This is for ElasticSearch Queries compatibility since a keyword field is always stored
+                            // by default when indexing without explicit mapping in ElasticSearch.
+                            // Keyword ignore above 256 chars by default.
+                            if (store == Field.Store.NO && stringValue.Length <= 256)
+                            {
+                                doc.Add(new StringField($"{entry.Name}.keyword", stringValue, Field.Store.YES));
                             }
```

Comment:
Here is the only change I've done in the Lucene implementation that truly matters.